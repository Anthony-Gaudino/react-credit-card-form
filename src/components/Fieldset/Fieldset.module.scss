.fieldset {
  padding: 0;
  margin: 0;
  border: none;

  &__legend {
    padding: 0;
    font-size: 0.8rem;
  }
}
