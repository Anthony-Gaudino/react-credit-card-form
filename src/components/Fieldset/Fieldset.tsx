import styles from "./Fieldset.module.scss";

interface FieldsetProps {
  /**
   * The fieldset legend text.
   */
  legend: string;
  /**
   * The fieldset children.
   */
  children: React.ReactNode;
}

/**
 * A HTML fieldset container with a legend.
 */
export const Fieldset = (props: FieldsetProps): JSX.Element => {
  const { legend, children } = props;

  return (
    <fieldset className={styles.fieldset}>
      <legend className={styles.fieldset__legend}>{legend}</legend>
      {children}
    </fieldset>
  );
};
