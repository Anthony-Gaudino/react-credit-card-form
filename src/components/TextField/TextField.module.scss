@use "../../theme.scss";

.textField {
  @include theme.field;
}
