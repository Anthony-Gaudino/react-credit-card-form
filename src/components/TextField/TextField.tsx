import clsx from "clsx";
import { FormEventHandler } from "react";

import InputLabel from "../InputLabel";
import styles from "./TextField.module.scss";

interface TextFieldProps {
  /**
   * Type of input field.
   */
  type: "email" | "number" | "password" | "search" | "tel" | "text" | "url";
  /**
   * Input field label.
   */
  label: string;
  /**
   * If `true` hides the label. Defaults to false;
   */
  hideLabel?: boolean;
  /**
   * Input field value.
   */
  value?: string;
  /**
   * Input field pattern regex.
   */
  pattern?: string;
  /**
   * Input field max string length.
   */
  maxLength?: number;
  /**
   * Input field max string length.
   */
  minLength?: number;
  /**
   * Type of input mode.
   */
  inputMode?:
    | "none"
    | "text"
    | "tel"
    | "url"
    | "email"
    | "numeric"
    | "decimal"
    | "search";
  /**
   * Input field max string length.
   */
  required?: boolean;
  /**
   * If `true` highlits the input if it's invalid.
   */
  highlightInvalid?: boolean;
  /**
   * onChange function callback.
   */
  onChange?: FormEventHandler;
  /**
   * onFocus function callback.
   */
  onFocus?: FormEventHandler;
}

/**
 * A HTML input with a label.
 */
export const TextField = (props: TextFieldProps): JSX.Element => {
  const {
    label,
    hideLabel = false,
    highlightInvalid = false,
    ...others
  } = props;

  const classes = clsx({
    [styles.textField]: true,
    [styles["textField--highlight-invalid"]]: highlightInvalid,
  });

  return (
    <InputLabel label={label} hidden={hideLabel}>
      <input className={classes} {...others}></input>
    </InputLabel>
  );
};
