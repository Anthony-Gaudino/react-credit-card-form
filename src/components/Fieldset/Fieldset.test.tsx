import { render, screen } from "@testing-library/react";

import TextField from "../TextField";
import Fieldset from ".";

describe("Test fieldset component", () => {
  test("the fieldset is correctly rendered", () => {
    const { asFragment } = render(
      <Fieldset legend="A legend">
        <TextField type="text" label="A label" />
      </Fieldset>
    );

    const fieldsetLegend = screen.getByText("A legend");
    const textFieldLabel = screen.getByText("A label");

    expect(fieldsetLegend).toBeInTheDocument();
    expect(textFieldLabel).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
