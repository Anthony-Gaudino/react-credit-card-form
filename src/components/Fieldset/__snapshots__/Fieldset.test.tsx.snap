// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test fieldset component the fieldset is correctly rendered 1`] = `
<DocumentFragment>
  <fieldset
    class="fieldset"
  >
    <legend
      class="fieldset__legend"
    >
      A legend
    </legend>
    <label
      class="inputLabel"
    >
      A label
      <input
        class="textField"
        type="text"
      />
    </label>
  </fieldset>
</DocumentFragment>
`;
